### ISSUE TITLE
Leduc public observer doesn't differentiate between pre and post chance node states

### ISSUE THREAD

**Issue by VitamintK** [deepmind/open_spiel#548]
Just a minor issue I noticed -- maybe it's even intended.

When making an observer for public info and no private info for Leduc, even if the observer is supposed to have perfect recall, the tensor immediately before and after a chance node seems to be exactly the same.

```
observer = make_observation(game, pyspiel.IIGObservationType(
            perfect_recall=True,
            public_info=True,
            private_info=pyspiel.PrivateInfoType.NONE))
observer.set_from(game.new_initial_state(), 0)
print(observer.tensor)
observer.set_from(game.new_initial_state().child(1), 0)
print(observer.tensor)
```
outputs
```
[1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
[1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
```

but I'd expect them to differ, or if calling `set_from()` on chance nodes is unintended, then I'd expect an error to be raised.

**Comment by lanctot** [comment]
@michalsustr @elkhrt I believe one of you changed this.. can one of you take a look?

**Comment by lanctot** [comment]
Haha no by recently I mean like 6-8 months ago. Bad word choice in retrospect.. :-p

I just meant one of them implemented the Leduc observers when they got changed and so they will be able to answer.

**Comment by lanctot** [comment]
Wait a sec.. is this the first chance node in the game? That is a private card deal. The public info should not change there. In Leduc, the only public chance node is at the start of round 2. 

Can you elaborate on why you expect them to be different?

**Comment by elkhrt** [comment]
Marc is right - nothing public has changed between the two states, unless you count the fact of a card having been dealt (which doesn't really seem worth recording?)

As always, I recommend looking at the playthroughs to understand the game dynamics, e.g.
https://github.com/deepmind/open_spiel/tree/master/open_spiel/integration_tests/playthroughs/leduc_poker_1540482260.txt

The playthrough doesn't contain the public observation tensor, but you can either look at the private observation (and remove the private card) or the public observation string which has the same information content as the tensor.

Do you see a potential problem with this approach?

**Comment by michalsustr** [comment]
I ran exactly into this issue and fixed it for myself locally, because it was breaking my algorithms :) I added `card_dealt` observation for each player to model "the passage of time".

If there is agreement on distinguishing the time for the chance nodes I can submit a PR.

**Comment by VitamintK** [comment]
> Can you elaborate on why you expect them to be different?

Yeah, I was running some code with the intention of finding all the states with the same public state.  I assumed that all states that had the same `tensor` were states within the same "public infoset". The scenario in the original post broke my assumptions, since when I used `game.new_initial_state().child(0).child(1)`, I assumed that the other states with the same `tensor` would also be post-hole-card-deals, but instead I also ran into states like `game.new_initial_state()` and `game.new_initial_state().child(0)`.  

Also what @michalsustr said about it being an intuitive public observation that the move has been made.

Maybe the current implementation is an optimal way to do it, since there are probably trade-offs of changing it, like larger public tensors.  But encoding "this public chance node has happened" in the public state, or raising an error when trying to set from a chance node would help users making the same assumption I did.

> If there is agreement on distinguishing the time for the chance nodes I can submit a PR.
Another thing I realized is that the same issue would apply to any move that doesn't affect the public state:  i.e. also players' secret moves (like the moves in phantom tic tac toe, or the first player's move in a simultaneous-turned-sequential game like rock-paper-scissors).  There are none in leduc, but I'd guess that for consistency you'd probably want to settle on a universal design decision.
